### PR TITLE
Add ability to retrieve prepared multi primes

### PR DESCRIPTION
**Pull request by adamdougal**
Also changed the other retrieve end points to return fixed delay.
Might also cover #122.

### CODE REVIEW

File: java-client/src/main/java/org/scassandra/http/client/types/GsonVariableMatchDeserialiser.java [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+import java.lang.reflect.Type;
+
+public class GsonVariableMatchDeserialiser implements JsonDeserializer<MultiPrimeRequest.VariableMatch> {
+    @Override
+    public MultiPrimeRequest.VariableMatch deserialize(JsonElement json, Type typeOfT, JsonDeserializationContext context) throws JsonParseException {
+        JsonObject jsonObject = (JsonObject) json;
+
+        String type = jsonObject.get("type").getAsString();
+
+        if(type.equalsIgnoreCase("any")) {
```
Comment (chbatey) [inline java-client/src/main/java/org/scassandra/http/client/types/GsonVariableMatchDeserialiser.java]: whitespace after the if

File: java-client/src/main/java/org/scassandra/http/client/types/GsonVariableMatchDeserialiser.java [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+import org.scassandra.http.client.MultiPrimeRequest;
+
+import java.lang.reflect.Type;
+
+public class GsonVariableMatchDeserialiser implements JsonDeserializer<MultiPrimeRequest.VariableMatch> {
+    @Override
+    public MultiPrimeRequest.VariableMatch deserialize(JsonElement json, Type typeOfT, JsonDeserializationContext context) throws JsonParseException {
+        JsonObject jsonObject = (JsonObject) json;
+
+        String type = jsonObject.get("type").getAsString();
```
Comment (chbatey) [inline java-client/src/main/java/org/scassandra/http/client/types/GsonVariableMatchDeserialiser.java]: Is this always serialised as a string?

File: java-client/src/test/java/org/scassandra/http/client/PrimingClientTest.java [hunk 3]
```diff
@@ -767,6 +755,48 @@ public void testDeletingOfPreparedMultiPrimesFailed() {
...
+        MultiPrimeRequest pr = MultiPrimeRequest.request()
+                .withWhen(when()
+                        .withQuery("select * from person where name = ?"))
+                .withThen(then()
+                        .withVariableTypes(TEXT)
+                        .withOutcomes(outcome(match().withVariableMatchers(exactMatch().withMatcher("Chris").build()), action()))
+                )
+                .build();
+
+        stubFor(get(urlEqualTo(PRIME_PREPARED_MULTI_PATH)).willReturn(aResponse().withStatus(200).withBody(
```
Comment (chbatey) [inline java-client/src/test/java/org/scassandra/http/client/PrimingClientTest.java]: We should test some other variable types

File: server/src/main/scala/org/scassandra/server/priming/prepared/PrimePreparedMultiStore.scala [hunk 4]
```diff
@@ -5,19 +5,17 @@ import java.util.concurrent.TimeUnit
...
 import org.scassandra.server.priming.routes.PrimingJsonHelper
 
 import scala.concurrent.duration.FiniteDuration
 
-class PrimePreparedMultiStore extends PreparedStoreLookup with LazyLogging {
-
-  var state: Map[PrimeCriteria, PreparedPrimeResult] = Map()
+class PrimePreparedMultiStore extends LazyLogging with PreparedStore[PrimePreparedMulti, PreparedMultiPrime] with PreparedStoreLookup {
 
   // todo change to a result to include validation failures etc
```
Comment (chbatey) [inline server/src/main/scala/org/scassandra/server/priming/prepared/PrimePreparedMultiStore.scala]: Update comment, still haven't dealt with validation failures tho

File: server/src/main/scala/org/scassandra/server/priming/prepared/PrimePreparedPatternStore.scala [hunk 5]
```diff
@@ -21,12 +21,12 @@ import com.typesafe.scalalogging.LazyLogging
...
 import org.scassandra.server.priming.query.{Prime, PrimeCriteria, PrimeMatch}
 import org.scassandra.server.priming.routes.PrimingJsonHelper
-import org.scassandra.server.priming.{Defaulter, PrimeAddResult, PrimeAddSuccess}
+import org.scassandra.server.priming.{PrimeAddResult, Defaulter, PrimeAddSuccess}
 import org.scassandra.server.priming.json.Success
 
 import scala.concurrent.duration.FiniteDuration
 
-class PrimePreparedPatternStore extends LazyLogging with PreparedStore with PreparedStoreLookup {
+class PrimePreparedPatternStore extends LazyLogging with PreparedStore[PrimePreparedSingle, PreparedPrime] with PreparedStoreLookup {
```
Comment (chbatey) [inline server/src/main/scala/org/scassandra/server/priming/prepared/PrimePreparedPatternStore.scala]: Switch LazyLogging and PreparedStore, doesn't make any difference but is a more accurate description of the type

File: server/src/main/scala/org/scassandra/server/priming/prepared/PrimePreparedStore.scala [hunk 6]
```diff
@@ -27,11 +27,11 @@ import org.scassandra.server.priming.routes.PrimingJsonHelper
 
 import scala.concurrent.duration.FiniteDuration
 
-class PrimePreparedStore extends LazyLogging with PreparedStore with PreparedStoreLookup {
+class PrimePreparedStore extends LazyLogging with PreparedStore[PrimePreparedSingle, PreparedPrime] with PreparedStoreLookup {
```
Comment (chbatey) [inline server/src/main/scala/org/scassandra/server/priming/prepared/PrimePreparedStore.scala]: Same

File: server/src/main/scala/org/scassandra/server/priming/routes/PrimingPreparedRoute.scala [hunk 7]
```diff
@@ -89,25 +121,8 @@ trait PrimingPreparedRoute extends HttpService with LazyLogging {
...
-              case _: SyntaxErrorResult => SyntaxError
-              case _: UnauthorizedResult => Unauthorized
-              case _: InvalidResult => Invalid
-              case _: ConfigErrorResult => ConfigError
-              case _: AlreadyExistsResult => AlreadyExists
-              case _: UnpreparedResult => Unprepared
-              case _: ClosedConnectionResult => ClosedConnection
-            }
+            val fixedDelay = if (preparedPrime.prime.fixedDelay.isDefined) Some(preparedPrime.prime.fixedDelay.get.toMillis) else None
+            val result = PrimingJsonHelper.convertToResultJsonRepresentation(preparedPrime.getPrime().result)
```
Comment (chbatey) [inline server/src/main/scala/org/scassandra/server/priming/routes/PrimingPreparedRoute.scala]: nice

File: server/src/main/scala/org/scassandra/server/priming/routes/PrimingPreparedRoute.scala [hunk 8]
```diff
@@ -53,6 +54,37 @@ trait PrimingPreparedRoute extends HttpService with LazyLogging {
...
           StatusCodes.OK
         }
+      } ~
+      get {
+        complete {
+
+          val preparedPrimes: Iterable[PrimePreparedMulti] = primePreparedMultiStore.retrievePrimes().map({ case (primeCriteria, preparedPrime) =>
+
+            val outcomes : List[Outcome] = preparedPrime.variableMatchers.map({ case (outcome) =>
+              val variableMatchers: List[VariableMatch] = outcome._1
```
Comment (chbatey) [inline server/src/main/scala/org/scassandra/server/priming/routes/PrimingPreparedRoute.scala]: Can extract from a tuple like:

```scala
val (variableMatchers, prime) = outcome
```

File: server/src/main/scala/org/scassandra/server/priming/routes/PrimingPreparedRoute.scala [hunk 9]
```diff
@@ -53,6 +54,37 @@ trait PrimingPreparedRoute extends HttpService with LazyLogging {
...
+              val criteria = Criteria(variableMatchers)
+
+              val prime: Prime = outcome._2
+              val result = PrimingJsonHelper.convertToResultJsonRepresentation(prime.result)
+              val fixedDelay = if (prime.fixedDelay.isDefined) Some(prime.fixedDelay.get.toMillis) else None
```
Comment (chbatey) [inline server/src/main/scala/org/scassandra/server/priming/routes/PrimingPreparedRoute.scala]: Should never need to call `isDefined`

```scala
prime.fixedDelay.map(_.toMillis)
```

File: server/src/main/scala/org/scassandra/server/priming/routes/PrimingPreparedRoute.scala [hunk 10]
```diff
@@ -89,25 +121,8 @@ trait PrimingPreparedRoute extends HttpService with LazyLogging {
...
-              case _: TruncateErrorResult => TruncateError
-              case _: SyntaxErrorResult => SyntaxError
-              case _: UnauthorizedResult => Unauthorized
-              case _: InvalidResult => Invalid
-              case _: ConfigErrorResult => ConfigError
-              case _: AlreadyExistsResult => AlreadyExists
-              case _: UnpreparedResult => Unprepared
-              case _: ClosedConnectionResult => ClosedConnection
-            }
+            val fixedDelay = if (preparedPrime.prime.fixedDelay.isDefined) Some(preparedPrime.prime.fixedDelay.get.toMillis) else None
```
Comment (chbatey) [inline server/src/main/scala/org/scassandra/server/priming/routes/PrimingPreparedRoute.scala]: Same as above

**Comment by chbatey** [comment]
This looks really good, just a few minor comments. Once you've addressed them could you please squash your commits.

File: java-client/src/main/java/org/scassandra/http/client/types/GsonVariableMatchDeserialiser.java (hunk 2 above)
Comment (adamdougal) [inline java-client/src/main/java/org/scassandra/http/client/types/GsonVariableMatchDeserialiser.java]: I believe so

**Comment by adamdougal** [comment]
Commits squashed

**Comment by chbatey** [comment]
Looks good

**Pull request merged**